libraries/AP_Filesystem/AP_Filesystem_Sys.cpp

```diff
@@ -158,8 +165,13 @@ int32_t AP_Filesystem_Sys::read(int fd, void *buf, uint32_t count)
         return -1;
     }
     struct rfile &r = file[fd];
-    count = MIN(count, r.str->get_length() - r.file_ofs);
-    memcpy(buf, &r.str->get_string()[r.file_ofs], count);
+    if (r.is_storage_bin) {
+        count = MIN(count, HAL_STORAGE_SIZE - r.file_ofs);
+        hal.storage->read_block((uint8_t *)buf, r.file_ofs, count);
```

Comment:
much simpler to just have a hal.storage function to return a const uint8_t * for storage. All the HALs have a memory backed copy

Reply:
yes, read_block does memcpy from that memory backed copy.